financial-connections/src/main/java/com/stripe/android/financialconnections/browser/CustomTabsManager.kt

```diff
@@ -0,0 +1,182 @@
+import androidx.browser.customtabs.CustomTabsServiceConnection
+import androidx.browser.customtabs.CustomTabsSession
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import com.stripe.android.core.Logger
+import javax.inject.Inject
+
+internal interface CustomTabsManager : DefaultLifecycleObserver {
+
+    /**
+     * Opens the URL on a Custom Tab if possible. Otherwise fallsback to opening it on a WebView.
+     *
+     * @param activity The host activity.
+     * @param uri the Uri to be opened.
+     * @param fallback a CustomTabFallback to be used if Custom Tabs is not available.
+     */
+    fun openCustomTab(
+        activity: Activity,
+        uri: Uri,
+        fallback: (Uri) -> Unit = {
+            activity.startActivity(Intent(Intent.ACTION_VIEW, uri))
+        }
+    )
+
+    /**
+     * Warms up the browser for a given URL, so that it loads faster when launched.
+     */
+    fun mayLaunchUrl(url: String): Boolean
+}
+
+/**
+ * Manages the connection to the CustomTabsService.
+ *
+ */
+internal class CustomTabsManagerImpl @Inject constructor(
+    private val logger: Logger
+) : CustomTabsManager {
+
+    private var client: CustomTabsClient? = null
+    private var connection: CustomTabsServiceConnection? = null
+    private var session: CustomTabsSession? = null
+
+    /**
+     * Binds the Activity to the CustomTabsService.
+     */
+    override fun onStart(owner: LifecycleOwner) {
+        super.onStart(owner)
+        val context = owner as? Context ?: return
+        if (client != null) return
+        connection = object : CustomTabsServiceConnection() {
+            override fun onCustomTabsServiceConnected(
+                name: ComponentName,
+                client: CustomTabsClient
+            ) {
+                log("Service connected")
+                this@CustomTabsManagerImpl.client = client
+                this@CustomTabsManagerImpl.client?.warmup(0)
+                session = this@CustomTabsManagerImpl.client?.newSession(CustomTabsCallback())
+            }
+
+            override fun onServiceDisconnected(name: ComponentName) {
+                log("Service disconnected")
+                client = null
+                session = null
+            }
+        }
+        bindCustomTabService(context)
+    }
+
+    /**
+     * Unbinds the Activity from the CustomTabsService.
+     */
+    override fun onStop(owner: LifecycleOwner) {
+        super.onStop(owner)
+        val context = owner as? Context ?: return
+        if (connection == null) return
+        connection?.let {
+            runCatching { context.unbindService(it) }
+                .onFailure { log("OnStop: couldn't unbind, ${it.stackTraceToString()}") }
+                .onSuccess { log("OnStop: service unbound") }
+        }
+        client = null
+        connection = null
+        session = null
+    }
+
+    /**
+     * Creates or retrieves an exiting CustomTabsSession.
```

Comment:
nit: It seems like the implementation returns either null or a new session, is that what we want? The docs for this function is misleading.